### PR TITLE
initial version of crop.py interface

### PR DESCRIPTION
**Pull request by thejohnhoffer**
Four functions assist with cropping arbitrary subregions.

### CODE REVIEW

**Comment by dpwrussell** [comment]
@thejohnhoffer Can you link the working example?

**Comment by thejohnhoffer** [comment]
@dpwrussell Here is [the gist](https://gist.github.com/thejohnhoffer/6539b0d39bf445f82b54cf0fcca72e96).

**Comment by thejohnhoffer** [comment]
Here is how the gist implements the `render_scaled_region` API of OMERO.figure using the `crop` module created in this pull request. 
```python
def assign_image(tile):

    c = tile['channel']
    i, j = tile['indices']

    # No negative tiles
    if i < 0 or j < 0:
        return None

    image = load_tile(c, k_lod, i, j)

    tile['image'] = image
    return tile

tiles = crop.iterate_tiles(channels, tile_size,
                           k_origin, k_shape)
return crop.stitch_tiles(map(assign_image, tiles),
                         tile_size, k_shape, order)
```

**Comment by thejohnhoffer** [comment]
### The implementation

I designed this module to flexibly handle the need to composite channels before or after stitching tiles. In the code, I use the word "tile" to refer to an image defined uniquely by x,y indices and channel id. So, `crop.iterate_tiles` yields a number of tiles equal to the product of x indices, y indices, and channel ids needed to meet a request. This allows `crop.stitch_tiles` to either hash these tiles by channel ids or by x,y indices.

Within `crop.stitch_tiles`, the `hash_groups` function creates a dictionary with each value storing a `Group` of tiles with either the same x,y indices or the same channel ids, depending on the desired order of the compositing and stitching operations. The `combine_groups` function then reduces this dictionary of groups of tiles into a single composited and stitched image. 

Regardless of the order of operations, `combine_groups` first merges all tiles in a group, then it merges a given group to the final output. If compositing 'before' stitching, a composited image for each group is stitched to the final output. If compositing 'after' stitching, a stitched image for each group is composited to the final output.

**Comment by dpwrussell** [comment]
@thejohnhoffer Can you update the Gist to have a `requirements.txt` which points at this branch of `minerva-lib-python`, it's hard to follow what is the example and what is the actual code.

**Comment by thejohnhoffer** [comment]
@dpwrussell 

Okay, I've updated the [gist and documentation](https://gist.github.com/thejohnhoffer/6539b0d39bf445f82b54cf0fcca72e96#gistcomment-2633588) It uses an external requirements.txt file, and it uses two functions added to minerva-lib-python to simplify the interface.

Now, the library user always works in full-resolution coordinates. The level of detail must now be passed to each library function so that the functions can calculate the down-sampled coordinates. 

- `crop.get_lod_1tile` gives the level of detail needed to return at most a full tile
```python
    level = crop.get_lod_1tile(levels, tile_size, *shape)
```
- `crop.get_lod_4tiles` gives the level of detail needed to return at most four tiles
```python
    level = crop.get_lod_4tiles(levels, tile_size, *shape)
```
- `crop.get_lod` gives the level of detail needed to return at most the `max_size`
```python
    level = crop.get_lod(levels, max_size, *shape)
```

After computing the level, here is the revised library usage:
```python
    def assign_image(tile):

        lod = tile['level']
        c = tile['channel']
        i, j = tile['indices']

        # Disallow negative tiles
        if i < 0 or j < 0:
            return None

        # Load image from Omero
        image = load_tile(c, lod, i, j)

        # Disallow empty images
        if image is None:
            return None

        # Return image to Minerva
        tile['image'] = image
        return tile

    # Load and stitch all tiles
    tiles = crop.list_tiles_at_level(channels, tile_size,
                                     origin, shape, level)
    image_tiles = map(assign_image, tiles)
    return crop.stitch_tiles_at_level(image_tiles, tile_size,
                                      shape, level, order)
```

**Comment by thejohnhoffer** [comment]
Alternatively, we could create new functions such as: 

```python
list_tiles_max_resolution(channels, tile_size, origin, shape, max_size)
stitch_tiles_max_resolution(image_tiles, tile_size, shape, max_size, order)
```

Then, the user could define `max_size = 4*min(*tile_size)` in order to restrict the output image to at most 2 x 2 tiles.

**Comment by dpwrussell** [comment]
I don't understand the 1-tile/4-tile thing. If the user wants to get a region that spans, for example, 6 tiles at a higher resolution, then that should totally be possible.

There should be a function `get_optimum_pyramid_level` (this looks like maybe `get_lod`, please try and match up the terminology with what we are already using) that calculates the optimum (i.e. the pyramid level that is closest to, but above, the request resolution). That function will simply return the pyramid level that should be used.

Next, a function should output the list of tile indices required for the given region (looks like `select_tiles`).

The calling function will then be in charge of constructing URLs and fetching them.

We then need to call a function which takes the resultant tiles does the required stitching and cropping (Looks like this is ). `iterate_tiles` is a terrible name, the fact that it is iterating over tiles is irrelevant to what it is actually doing. Same goes for `list_tiles_at_level`.

Can we also add more explanation as to what all of these functions are doing. In every case I had to read the code in order to determine what they do.

The whole groups thing is still very confusing to me, I can't see any reason why we'd want to stitch before compositing so if it's just to add that capability, then I don't think it's worth it.

**Comment by thejohnhoffer** [comment]
Okay, I updated the crop.py library functions to behave how @dpwrussell described. Here is the calling function:


```python
def do_crop(load_tile, channels, tile_size, full_origin, full_size,
            levels=1, max_size=2000):
    ''' Interface with minerva_lib.crop
    Args:
        load_tile: Function to supply 2D numpy array
        channels: List of dicts of channel rendering settings
        tile_size: The width, height of a single tile
        full_origin: Request's full-resolution x, y origin
        full_size: Request's full-resolution width, height
        levels: The number of pyramid levels
        max_size: The maximum response width or height
    Returns:
        2D numpy float array of with width, height given by
        `full_size` if `full_size <= max_size` or width, height given by
        `full_size / 2 ** l <= max_size` for the lowest `l` meeting
        `l < levels`. The array is a composite of all channels
        for full or partial tiles within `full_size` from `full_origin`.
    '''

    level = crop.get_optimum_pyramid_level(full_size, levels, max_size)
    crop_origin = crop.scale_by_pyramid_level(full_origin, level)
    crop_size = crop.scale_by_pyramid_level(full_size, level)
    print(f'Cropping 1/{level} scale')

    image_tiles = []

    for channel in channels:

        (red, green, blue) = channel['color']
        _id = channel['channel']
        _min = channel['min']
        _max = channel['max']

        for indices in crop.select_tiles(tile_size, crop_origin, crop_size):

            (i, j) = indices

            # Disallow negative tiles
            if i < 0 or j < 0:
                continue

            # Load image from Omero
            image = load_tile(_id, level, i, j)

            # Disallow empty images
            if image is None:
                continue

            # Add to list of tiles
            image_tiles.append({
                'min': _min,
                'max': _max,
                'channel': _id,
                'image': image,
                'indices': (i, j),
                'color': (red, green, blue),
            })

    return crop.stitch_tiles(image_tiles, tile_size, crop_origin, crop_size)
```

Here is what it does:
  - It calls `crop.get_optimum_pyramid_level` to get `level`
  - It calls `crop.scale_by_pyramid_level` to scale the origin and size
  - It gets all the tile indices from `crop.select_tiles`
  - It loads all images into `image_tiles`
  - It stitches and composites with `crop.stitch_tiles`


---

I'd like to make sure I understand this:

> "the pyramid level that is closest to, but above, the request resolution." 

It's possible this describes how I've implemented `get_optimum_pyramid_level`, but I want to be certain. I've implemented this function to replicate the `render_scaled_region` endpoint called by OMERO.figure.

When OMERO.figure calls the `render_scaled_region` endpoint, it always supplies coordinates measured at pyramid level 0, meaning coordinates at full resolution. In one example, it asks for 48960 x 48960 pixels (implicitly at pyramid level 0). In order to keep the resulting image below 2000 x 2000 pixels, OMERO.web returns  1530 x 1530 pixels from pyramid level 5.

The function used by `render_scaled_region` uses the width, height in the first column to return the pyramid level in the second column. Then that pyramid level transforms the level-0 coordinates to return the width and height in the third column:

| level-0 width, height | resulting pyramid level | resulting width, height |
|---------------|-------------------------|----------------------|
| 2000, 2000    | 0                       | 2000, 2000           |
| 2001, 2001    | 1                       | 1000, 1000           |
| 4000, 4000    | 1                       | 2000, 2000           |
| 4001, 4001    | 2                       | 1000, 1000           |

I have implemented this functionality as `get_optimum_pyramid_level` in minerva-lib-python. It requires 3 arguments:
- the coordinates measured within pyramid level 0
- the number of levels of detail
- the maximum
  - this is 2000 for the OMERO.web `render_scaled_region` endpoint
  - It could be N*1024 to ensure at most N**2 tiles

**Comment by thejohnhoffer** [comment]
Though the tests passed, the build is failing...
```
The command "python setup.py test" exited with 0.
Done. Your build exited with 1.
```

**Comment by thejohnhoffer** [comment]
select_tiles should throw exceptions for negative coordinates and coordinates outside bounds
stitch_tiles / stitch_tile should throw similar exceptions for indices that are out of bounds

File: src/minerva_lib/crop.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    Returns:
+        A numpy array with the resized `source`
+    '''
+
+    if factor <= 0:
+        raise ValueError('Scale factor must be above zero')
+
+    factors = [factor, factor, 1]
+
+    out_shape = [int(round(a * b)) for a, b in zip(source.shape, factors)]
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: This is needed such that the output image has the same number of color channels as the input image.

File: src/minerva_lib/crop.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        The part of the tile within the requested region
+        Subregion start, end relative to origin of tile
+    '''
+
+    tile_start = np.int64(indices) * tile_size
+    crop_end = np.int64(crop_origin) + crop_size
+    tile_end = tile_start + tile_size
+
+    # Should be true that 0 <= start_uv < end_uv <= tile_size
+    start_uv = np.maximum(crop_origin, tile_start) - tile_start
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: In x, or y at starts at the start of the tile or the start of the requested region.

File: src/minerva_lib/crop.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    end_uv = np.minimum(tile_end, crop_end) - tile_start
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: In x, or y at ends at the end of the tile or the end of the requested region.

File: src/minerva_lib/crop.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        crop_origin: x, y origin of requested image region
+
+    Returns:
+        Position of tile region within requested image region
+    '''
+
+    tile_start = np.int64(indices) * tile_size
+
+    # Should never be negative
+    return np.maximum(crop_origin, tile_start) - crop_origin
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: The position is at the start of the tile or the start of the requested region.

File: src/minerva_lib/crop.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+        crop_size: width, height of requested image region
+
+    Returns:
+        For a given `shape` of `(width, height)`,
+        returns a float32 RGB color image with shape
+        `(height, width, 3)` and values in the range 0 to 1
+    '''
+    inputs = [t for t in tiles if t]
+    rgb_tile_size = tuple(tile_size) + (3,)
+    stitched = np.zeros(tuple(crop_size) + (3,))
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: This is the final output, which is a 3-channel RGB array.

File: src/minerva_lib/crop.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+            out_x = min(int(round(source_x * factor)), out_max_x)
+            out_y = min(int(round(source_y * factor)), out_max_y)
+            out[out_y, out_x] = source[source_y, source_x]
+
+    return out
+
+
+def get_optimum_pyramid_level(input_shape, level_count,
+                              max_size=None, round_up=True):
+    ''' Gives a pyramid level that scales a region near a maximum size.
```
Comment (dpwrussell) [inline src/minerva_lib/crop.py]: I don't understand this description. I think a better summary would be something like.

"Calculates the optimum pyramid level to use rendering an image region into an output image with a given bound".

File: tests/crop.py [hunk 7]
```diff
@@ -0,0 +1,907 @@
...
+    return [256, 256]
+
+
+@pytest.fixture
+def real_shape_1024x1024():
+    return [1024, 1024]
+
+
+@pytest.fixture
+def real_stitched_with_gamma(dirname):
```
Comment (dpwrussell) [inline tests/crop.py]: How big are the test images in S3? I think if possible it would be preferable not to need internet access to run the tests.

File: src/minerva_lib/crop.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    that is required for the requested image region.
+
+    Args:
+        indices: integer i, j tile indices
+        tile_size: width, height of one tile
+        crop_origin: x, y origin of requested image region
+        crop_size: width, height of requested image region
+
+    Returns:
+        The part of the tile within the requested region
```
[the remark this answers is not in the document]
Comment (thejohnhoffer) [reply src/minerva_lib/crop.py]: I believe I mean the above line to replace the below line. Following your earlier suggestions, the above line gives a higher-level description than the below line.

File: src/minerva_lib/crop.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        return False
+
+    return True
+
+
+def select_tiles(tile_size, crop_origin, crop_size):
+    ''' Selects tile indices to fill the requested region.
+
+    Args:
+        tile_size: width, height of one tile
```
[the remark this answers is not in the document]
Comment (thejohnhoffer) [reply src/minerva_lib/crop.py]: It could be a tuple, list, or numpy array.  In the tests, I use a list. I can document and test these as any type you prefer. In the `blend` module, we use numpy arrays for arguments like color and range. For the `blend` module, I documented "float array" to emphasize that the "color" argument should have floating-point values, but it is true that the "color" argument could be a list, tuple, or array. In this new crop module, the documentation could explicitly ask for arrays of two integers, lists of two integers, or tuples of two integers. This is true of all the coordinate arguments in this module.

File: src/minerva_lib/crop.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+        crop_origin: x, y origin of requested image region
+        crop_size: width, height of requested image region
+
+    Returns:
+        For a given `shape` of `(width, height)`,
+        returns a float32 RGB color image with shape
+        `(height, width, 3)` and values in the range 0 to 1
+    '''
+    inputs = [t for t in tiles if t]
```
Comment (thejohnhoffer) [inline src/minerva_lib/crop.py]: check if needed

**Comment by dpwrussell** [comment]
Closing in favour of #23
